pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,73 @@
+from typing import List, cast
```

Comment:
@algoidurovic Minor and looking to test if there's a convention we wish to uphold:  When defining packages, do we try to avoid circular dependencies?

* I'm noting `pyteal/compiler/compiler.py` refers to a _deeper_ package:  `pyteal/compiler/optimizer/optimizer.py`.
* Typically, to minimize the chance for a circular dependency forming, I'd aim to keep communication flowing from less-nested-to-more-deeply-nested packages.
* In this case, _if_ we wish to enforce a dependency flow, possible actions include removing the `optimizer` package or moving `compiler` to a subpackage.
* I'll follow the group's convention.  I raise the question because I've had experiences where later refactorings can be complicated by circular dependencies.

Reply:
Quoting Jason:
"The approach PyTeal currently uses is below. Here a module means a directory of source code where all objects meant for external use are exported by the __init__.py source file in that directory.
When importing an object from the same module or a parent module, import the full path of the source file that defines the object.
When importing an object from a child module, import the entire module folder and access the desired object from there. Do not directly import the file that defines the object.
With this approach, circular dependencies can happen in two ways:
**Intra-module circular dependencies**, e.g. m1/a.py imports m1/b.py which imports m1/a.py. When this happens, normally one of the files only needs to import the other to implement something, so the import statements can be moved from the top of the file to the body of the function that needs them. This breaks the import cycle.
**Inter-module circular dependencies**, e.g. m1/a.py imports m1/m2/__init__.py which imports m1/m2/x.py which imports m1/a.py. To avoid this, make sure that any objects/files in m1 that need to be exposed to deeper modules do not rely on any objects from that deeper module. If that isn’t possible, then perhaps m1/a.py belongs in the m1/m2/ module"

Reply:
@algoidurovic Unresolving for a clarification:  

From the above note, I understood `pyteal/compiler/compiler.py` is liable to create _inter-module circular dependencies_ with `pyteal/compiler/optimizer/optimizer.py`.   Does that match your understanding?

If so, are we leaving the package structure unchanged because no issue exists today or was the thread unintentionally resolved?

Reply:
Summarizing a verbal with @algoidurovic that resolves the thread:  

We agreed there's no strong enforcement for circular dependencies today.  Since no circular dependency exists in the PR, we agreed to leave the structure as is.